repo/upgrade_lock.go

```diff
@@ -0,0 +1,123 @@
+package repo
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+
+	"github.com/kopia/kopia/repo/content"
+)
+
+func (r *directRepository) updateRepoConfig(ctx context.Context, cb func(repoConfig *repositoryObjectFormat) error) (*repositoryObjectFormat, error) {
+	f := r.formatBlob
+
+	repoConfig, err := f.decryptFormatBytes(r.formatEncryptionKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to decrypt repository config")
+	}
+
+	if err := cb(repoConfig); err != nil {
+		return nil, err
+	}
+
+	if err := encryptFormatBytes(f, repoConfig, r.formatEncryptionKey, f.UniqueID); err != nil {
+		return nil, errors.Errorf("unable to encrypt format bytes")
+	}
+
+	if err := writeFormatBlob(ctx, r.blobs, f, r.blobCfgBlob); err != nil {
+		return nil, errors.Wrap(err, "unable to write format blob")
+	}
+
+	if cd := r.cachingOptions.CacheDirectory; cd != "" {
+		if err := os.Remove(filepath.Join(cd, FormatBlobID)); err != nil && !os.IsNotExist(err) {
+			return nil, errors.Errorf("unable to remove cached repository format blob: %v", err)
+		}
+	}
+
+	return repoConfig, nil
+}
+
+// SetUpgradeLockIntent sets the upgrade lock intent on the repository format
+// blob for other clients to notice. If a lock intent was already placed then
+// it updates the existing lock using the output of the UpgradeLock.Update().
+//
+// This method also backs up the original format version on the upgrade lock
+// intent and sets the latest format-version o nthe repository blob. This
+// should cause the unsupporting clients (non-upgrade capable) to fail
+// connecting to the repository.
+func (r *directRepository) SetUpgradeLockIntent(ctx context.Context, l content.UpgradeLock) (*content.UpgradeLock, error) {
+	repoConfig, err := r.updateRepoConfig(ctx, func(repoConfig *repositoryObjectFormat) error {
+		if err := l.Validate(); err != nil {
+			return errors.Wrap(err, "invalid upgrade lock intent")
+		}
+
+		if repoConfig.FormattingOptions.UpgradeLock == nil {
+			// when we are putting a new lock then ensure that we can upgrade
+			// to that version
+			if repoConfig.FormattingOptions.Version >= content.MaxFormatVersion {
+				return errors.Errorf("repository is already upgraded to %d, we can only upgrade till %d",
```

Comment:
Something like this maybe?
```suggestion
				return errors.Errorf("repository is using version %d, and version %d is the maximum",
```

Reply:
you'd need to update the unit test if you choose to reword this

Reply:
Will do.